### PR TITLE
Reimplement integer length

### PR DESCRIPTION
**Pull request by chrisseaton**
https://twitter.com/ChrisGSeaton/status/1398459785092153346

Not as good as I was expecting. Seems to warm up a lot quicker though. What do you think?

```
Warming up --------------------------------------
              length   913.032k i/100ms
Calculating -------------------------------------
              length     51.007M (± 8.1%) i/s -    252.910M in   4.997260s

Warming up --------------------------------------
              length     5.735M i/100ms
Calculating -------------------------------------
              length     57.091M (± 2.8%) i/s -    286.766M in   5.027562s
```

### CODE REVIEW

**Comment by eregon** [comment]
If we can have a node for it, we could have an inline cache specialization (with a fairly small limit like 1, 2 or 3) which guards against the previous and next powers of ten, and just return a constant for the length.

For Time formatting though it's usually fixed-width zero-padded anyway though, so it can probably be simpler:
```
> Time.now.strftime "%m"
=> "05"
```
Probably we'd want a variant of LazyIntRope like ZeroPaddedLazyIntRope or so.

**Comment by eregon** [comment]
Could you show the benchmark to see which inputs you considered?

**Comment by chrisseaton** [comment]
```ruby
require 'benchmark/ips'

puts RUBY_DESCRIPTION

n = 0

Benchmark.ips do |x|
  x.report('length') do
    n = rand(1000000000).to_s.length
  end
end
```

**Comment by chrisseaton** [comment]
> For Time formatting though it's usually fixed-width zero-padded anyway though, so it can probably be simpler:

The padded numbers in time formatting are in `RopeConstants` so are just normal leaf ropes. The year has a fixed constant length.

https://github.com/oracle/truffleruby/pull/2361/files#diff-2165f1135ce1cc7280599d50b87438aee142f40977cc937dbfd175a8c79795fdR664

> Probably we'd want a variant of LazyIntRope like ZeroPaddedLazyIntRope or so.

Is there any cost to adding more rope types? I can think of lots to add - quickest win could be an n-way concat, for example. Can we add tens of them?

File: src/main/java/org/truffleruby/core/rope/LazyIntRope.java [hunk 1]
```diff
@@ -27,29 +27,43 @@ public LazyIntRope(int value) {
...
             sign = 0;
         }
 
-        return sign + (value < 1E5
-                ? value < 1E2 ? value < 1E1 ? 1 : 2 : value < 1E3 ? 3 : value < 1E4 ? 4 : 5
-                : value < 1E7 ? value < 1E6 ? 6 : 7 : value < 1E8 ? 8 : value < 1E9 ? 9 : 10);
+        final int bits = 31 - Integer.numberOfLeadingZeros(value);
+        int digits = ((77 * bits) >>> 8);
+
+        if (value > LENGTH_TABLE[digits]) {
```
Comment (franz1981) [inline src/main/java/org/truffleruby/core/rope/LazyIntRope.java:62]: I would create a larger array (16 length) with many zero values and masking digit in orse to hell the compiler to save the bound check while accessing the int[] static final array!

**Review by franz1981 — CHANGES_REQUESTED** [review]

File: src/main/java/org/truffleruby/core/rope/LazyIntRope.java (hunk 1 above)
Comment (franz1981) [reply src/main/java/org/truffleruby/core/rope/LazyIntRope.java:62]: I have used a similar trick here: https://github.com/franz1981/netty/blob/7c3513f12230bf0f46a48856922209934ba30987/common/src/main/java/io/netty/util/internal/StringUtil.java#L61

**Comment by eregon** [comment]
> The padded numbers in time formatting are in RopeConstants so are just normal leaf ropes.

Right, `RopeConstants.paddedNumber` seems great (I did not review that PR in details yet).

> Is there any cost to adding more rope types?
> I can think of lots to add - quickest win could be an n-way concat, for example. Can we add tens of them?

Yes, all nodes operating on specific Ropes must handle the new type.
Also, we are planning to migrate to TruffleString that becomes available (it's making good progress), and that will likely be harder if we have more Rope variants.
There is also of course a risk that some new Rope type isn't handled somewhere and that could cause bugs.
So I think we should only add what's really needed or very beneficial for performance.

File: src/main/java/org/truffleruby/core/rope/LazyIntRope.java (hunk 1 above)
Comment (franz1981) [reply src/main/java/org/truffleruby/core/rope/LazyIntRope.java:62]: And I would replace the if part with some branchless trick ;)
Going completely branchless and bound check is key to make this faster with unpredictable inputs

**Review by franz1981 — CHANGES_REQUESTED** [review]

**Comment by eregon** [comment]
The benchmark in https://github.com/oracle/truffleruby/pull/2365#issuecomment-850820680 might be dominated by the cost of `rand`. Could you try something like e.g.
```ruby
require 'benchmark/ips'

puts RUBY_DESCRIPTION

n = 0
numbers = 100.times.map { rand(1000000000) }

Benchmark.ips do |x|
  x.report('length') do
    numbers.each { |v| n = v.to_s.length }
  end
end
```

**Comment by chrisseaton** [comment]
```
ruby 2.7.3p183 (2021-04-05 revision 6847ee089d) [x86_64-darwin20]
Warming up --------------------------------------
              length     6.195k i/100ms
Calculating -------------------------------------
              length     62.706k (± 3.3%) i/s -    315.945k in   5.045198s

truffleruby 21.2.0-dev-7d950d5b, like ruby 2.7.3, GraalVM CE JVM [x86_64-darwin]
Warming up --------------------------------------
              length   651.096k i/100ms
Calculating -------------------------------------
              length      6.200M (± 6.0%) i/s -     31.253M in   5.062144s

truffleruby 21.2.0-dev-7d950d5b, like ruby 2.7.3, GraalVM CE JVM [x86_64-darwin]
Warming up --------------------------------------
              length   505.515k i/100ms
Calculating -------------------------------------
              length      6.151M (±11.5%) i/s -     30.331M in   5.001263s
```

Not statistically significant. Yeah probably not worth it then. Seems crazy there isn't a better way to do this...

**Comment by eregon** [comment]
I used a variant of the benchmark to try to measure more precisely:
```ruby
# truffleruby_primitives: true

require 'benchmark/ips'

puts RUBY_DESCRIPTION

small_numbers = [1, 9, 11, 99, 111, 999, 1111, 9999]
rand = Random.new(0)
random_numbers = 10.times.map { rand.rand(1000000000) }
p random_numbers

Benchmark.ips do |x|
  x.iterations = 2
  
  x.report('length small rand') do
    small_numbers.each { |v| Primitive.blackhole v.to_s.length }
  end
  
  x.report('length big rand') do
    random_numbers.each { |v| Primitive.blackhole v.to_s.length }
  end
end
```

Before:
```
   length small rand     68.602M (± 2.2%) i/s -    344.101M in   5.018525s
     length big rand     89.620M (± 0.1%) i/s -    456.744M in   5.096464s
   length small rand     68.804M (± 0.9%) i/s -    344.101M in   5.001580s
     length big rand     89.554M (± 0.1%) i/s -    447.788M in   5.000233s
```

This PR:
```
   length small rand     76.752M (± 0.1%) i/s -    390.447M in   5.087134s
     length big rand     65.355M (± 1.9%) i/s -    326.767M in   5.001952s
   length small rand     76.581M (± 1.3%) i/s -    390.447M in   5.099411s
     length big rand     65.494M (± 0.1%) i/s -    333.302M in   5.089015s
```
So small is slightly faster, big much slower.

Using the improved variant from https://github.com/lemire/Code-used-on-Daniel-Lemire-s-blog/blob/master/2021/05/28/digitcount.java#L7-L13 :
```
   length small rand     97.498M (± 0.3%) i/s -    495.377M in   5.080910s
     length big rand     72.909M (± 0.7%) i/s -    370.870M in   5.086982s
   length small rand     97.402M (± 0.9%) i/s -    495.377M in   5.086326s
     length big rand     72.779M (± 0.5%) i/s -    370.870M in   5.095968s

```
Small is much faster, big is still quite a bit slower.

I guess in practice it's more frequent to have smaller numbers so it might be a good trade-off, but unsure.
The code is obviously less clear.

---

Trying to remove the bound checks with `& 0xf` and making the array bigger doesn't to change the results:
```
   length small rand     97.363M (± 0.3%) i/s -    491.135M in   5.044447s
     length big rand     68.203M (± 4.1%) i/s -    341.794M in   5.020171s
   length small rand     92.409M (± 9.3%) i/s -    456.670M in   4.999764s
     length big rand     68.552M (± 4.3%) i/s -    348.496M in   5.094852s
vs
   length small rand     97.411M (± 0.3%) i/s -    495.123M in   5.082838s
     length big rand     69.195M (± 1.6%) i/s -    349.446M in   5.051373s
   length small rand     97.117M (± 2.0%) i/s -    485.414M in   5.000554s
     length big rand     68.464M (± 2.4%) i/s -    342.315M in   5.002983s
```

File: src/main/java/org/truffleruby/core/rope/LazyIntRope.java (hunk 1 above)
Comment (eregon) [reply src/main/java/org/truffleruby/core/rope/LazyIntRope.java:62]: I tried `& 0xf` but it doesn't to improve things.

> replace the if part with some branchless trick

How would you do that?

**Review by eregon — APPROVED** [review]

**Comment by eregon** [comment]
Unfortunately this does not work for `Integer.MIN_VALUE = -2147483648`, because `-value` is just the same as `value` and the increment after the lookup table does not work. Probably we need to handle it as a special case, similar to before.

**Comment by chrisseaton** [comment]
Possible new idea https://lemire.me/blog/2021/06/03/computing-the-number-of-digits-of-an-integer-even-faster/.

**Comment by eregon** [comment]
Great, that is quite a bit faster! I'll update to that.
```
   length small rand    116.362M (± 0.8%) i/s -    583.549M in   5.015233s
     length big rand     94.310M (± 1.9%) i/s -    473.265M in   5.020145s
   length small rand    116.494M (± 0.6%) i/s -    583.549M in   5.009469s
     length big rand     94.256M (± 2.7%) i/s -    473.265M in   5.025599s
```

So now it's 1.7x faster for small, and 1.05x faster for big than `Before`, and the first time it's faster than Before for big.

**Comment by chrisseaton** [comment]
Sorry, @eregon can you edit the change log to also credit @lemire?

**Comment by lemire** [comment]
Just in case this needs saying... the license on my blog is "public domain"... meaning that taking from the source code of my blog and doing "whatever" is encouraged.

**Comment by eregon** [comment]
Done, here is what I have (should merge soon): https://github.com/oracle/truffleruby/compare/master...eregon:faster-integer-length

**Pull request merged**